### DISCUSSION TITLE
Tutorial: Support non-Laravel apps LocalValetDriver

### DISCUSSION THREAD

**Discussion by cord** [beyondcode/herd-community#186] [Show and tell]
### Feature Description

Herd can run **non-laravel** apps with sub-directories, just place the following `LocalValetDriver.php` in your root directory of the test-site.

enjoy!

```php
<?php

namespace Valet\Drivers;


class LocalValetDriver extends ValetDriver
{
    /**
     * Determine if the driver serves the request.
     */
    public function serves(string $sitePath, string $siteName, string $uri): bool
    {
        return true;
    }

    /**
     * Determine if the incoming request is for a static file.
     */
    public function isStaticFile(string $sitePath, string $siteName, string $uri)/*: string|false */
    {
        $staticFilePath = $sitePath . $uri;

        // support index.html
        if (is_dir($staticFilePath)) {
            if (file_exists($f = str_replace('//', '/', $staticFilePath . '/index.html'))) {
                return $f;
            }
        }

        if (file_exists($staticFilePath) && !is_dir($staticFilePath)) {
            return $staticFilePath;
        }
        return false;
    }

    /**
     * Get the fully resolved path to the application's front controller.
     */
    public function frontControllerPath(string $sitePath, string $siteName, string $uri): string
    {
        $path = $sitePath . $uri;

        // The requested resource is a directory and contains a child ‘index.php’ file.
        if (file_exists($path . '/index.php')) {
            return $path . '/index.php';
        }

        // The requested resource is not a PHP file.
        if (file_exists($path) && isset(pathinfo($path)['extension']) && pathinfo($path)['extension'] != 'php') {
            return $path;
        }

        // The requested resource is a PHP file.
        if (file_exists($path) && isset(pathinfo($path)['extension']) && pathinfo($path)['extension'] == 'php') {
            return $path;
        }

        // 404
        return false;
    }
}
```

### Is this feature valuable for other users as well and why?

because herd is the way to go!

**Comment by thiskbj** [comment]
I'm getting a 404 error when I drop this in - any tips?

**Comment by cord** [comment]
@thiskbj what does the URL look like? you can try to debug with `echo  $sitePath. $uri; exit();`

**Comment by thiskbj** [comment]
Ah, that helped and it's working now, thank you! The URL looked fine, and my file looks exactly like what you posted, so I think the debugging just helped me understand how to use it.

In case this helps someone else, here's my use case: 

+ I have a Herd folder, pma - it's phpMyAdmin as downloaded, no Laravel anywhere in the folder.
+ LocalValetDriver.php is in the root
+ I updated config.inc.php to this:

[$cfg['Servers'][$i]['auth_type'] = 'cookie';
$cfg['Servers'][$i]['host'] = '127.0.0.1';
$cfg['Servers'][$i]['compress'] = false;
$cfg['Servers'][$i]['AllowNoPassword'] = true;](url)

I downloaded DBnign, started a mysql database, and now I can connect with my databases through phpMyAdmin. Now I can upload my databases for Herd to get started!

Thanks again!

**Comment by emilsundberg** [comment]
Just a heads up, after upgrading to Herd 1.2 I had to remove the namespace and extend LaravelValetDriver liket this:

```php
use Valet\Drivers\LaravelValetDriver;

class LocalValetDriver extends LaravelValetDriver {
     // Stuff
}
```

I found it in the official Laravel docs https://laravel.com/docs/10.x/valet#local-drivers since Herd 1.2 now supports custom Valet drivers properly according to the changelog (https://herd.laravel.com/changelog)

**Reply by kayintveen** [reply]
This indeed works. its not clear to me yet how to do this with a global custom driver though./

**Comment by CamKem** [comment]
@cord Do we actually need to set up a valet driver for a vanilla php site?
Trying to avoid using `php` homebrew install for basic vanilla websites I work on from time to time & would be nice just to use Herd.
I just created a folder & added a  /public/index.php and echoed a string and it works.
Will I run into problems unless I use a custom valet driver? If so, is there a standard driver we can use just for basic vanilla php sites?

**Comment by CristianSitov** [comment]
My Herd setup points to `~/Sites` folder.
I have my Laravel app in `~/Sites/backend` folder.
My SPA app built on Angular is in `~/Sites/frontend` folder.

I tried adding the LocalValetDriver in the SPA folder, but it states that:
```
Fatal error: Uncaught Error: Class "LocalValetDriver" not found in /Applications/Herd.app/Contents/Resources/valet/cli/Valet/Drivers/ValetDriver.php:55
Stack trace:
#0 /Applications/Herd.app/Contents/Resources/valet/server.php(54): Valet\Drivers\ValetDriver::assign('/Users/cristian...', 'frontend', '/')
#1 {main} thrown in /Applications/Herd.app/Contents/Resources/valet/cli/Valet/Drivers/ValetDriver.php on line 55
```

**Reply by alenny811** [reply]
Removing 

`namespace Valet\Drivers;`

from the beginning of the file seems to work
alongside with adding this:

```
use Valet\Drivers\LaravelValetDriver;

class LocalValetDriver extends LaravelValetDriver {
     // Stuff
}
```

**Comment by ac-steve** [comment]
For those of you wanting to use Herd to serve a simple JavaScript application, I was able to serve the production version of a simple JS app built with Vite using the following LocalValetDriver.

```php
use Valet\Drivers\ValetDriver;

class LocalValetDriver extends ValetDriver
{
    /**
     * Determine if the driver serves the request.
     */
    public function serves(string $sitePath, string $siteName, string $uri): bool
    {
        return true;
    }

    /**
     * Determine if the incoming request is for a static file.
     */
    public function isStaticFile(string $sitePath, string $siteName, string $uri)/*: string|false */
    {
        if (file_exists($staticFilePath = $sitePath . '/dist/' . $uri)) {
            return $staticFilePath;
        }

        return false;
    }

    /**
     * Get the fully resolved path to the application's front controller.
     */
    public function frontControllerPath(string $sitePath, string $siteName, string $uri): string
    {
        return $sitePath . '/dist/index.html';
    }
}
```
